src/centralized_telescope.rs

```diff
@@ -0,0 +1,421 @@
+        let mut bins: Vec<Vec<Element>> = Vec::with_capacity(setup.n_p as usize);
+        for _ in 0..setup.n_p {
+            bins.push(Vec::new());
+        }
+        // Take only up to 2*np elements for efficiency
+        for &s in set.iter().take(setup.n_p.saturating_mul(2) as usize) {
+            match Self::h0(setup, v, s) {
+                Some(h) => {
+                    bins[h as usize].push(s);
+                }
+                None => return (0, None),
+            }
+        }
+
+        let mut limit = 0;
+        for t in 0..setup.d {
+            if limit == setup.b {
+                return (limit, None);
+            }
+            if let Some(r) = Round::new(v, t, setup.n_p, setup.sec_param) {
+                let (l, proof_opt) = Self::dfs(setup, &bins, &r, limit.saturating_add(1));
+                if proof_opt.is_some() {
+                    return (l, proof_opt);
+                }
+                limit = l;
+            }
+        }
+        (limit, None)
+    }
+
+    /// Alba's proving algorithm, based on a depth-first search algorithm.
+    /// Calls up to setup.r times the prove_index function and returns an empty
+    /// proof if no suitable candidate is found.
+    pub fn prove(setup: &Setup, set: &[Element]) -> Option<Self> {
+        (0..setup.r).find_map(|v| Self::prove_index(setup, set, v).1)
+    }
+
+    /// Alba's verification algorithm, follows proving algorithm by running the
+    /// same depth-first search algorithm.
+    pub fn verify(setup: &Setup, proof: &Self) -> bool {
+        if proof.t >= setup.d || proof.v >= setup.r || proof.items.len() as u64 != setup.u {
+            return false;
+        }
+        if let Some(mut round) = Round::new(proof.v, proof.t, setup.n_p, setup.sec_param) {
+            for element in proof.items.clone() {
+                if let Some(h) = Self::h0(setup, proof.v, element) {
+                    if round.h_u64 == h {
+                        if let Some(r) = Round::update(&round, element, setup.sec_param) {
+                            round = r;
+                        } else {
+                            return false;
+                        }
+                    } else {
+                        return false;
+                    }
+                } else {
+                    return false;
+                }
+            }
+            return Self::h2(setup, &round);
+        }
+        false
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::test_utils::gen_items;
+    use rand_chacha::ChaCha20Rng;
+    use rand_core::{RngCore, SeedableRng};
+
+    #[test]
+    fn test_verify() {
+        let mut rng = ChaCha20Rng::from_seed(Default::default());
+        let nb_tests = 1_000;
+        let set_size = 1_000;
+        for _t in 0..nb_tests {
+            let seed = rng.next_u32().to_be_bytes().to_vec();
+            let s_p = gen_items::<DATA_LENGTH>(&seed, set_size);
+            let params = Params {
```

Comment:
maybe move it up to where the other constants are?

Reply:
Sure